packages/mdc-list/foundation.js

```diff
@@ -333,44 +309,197 @@ class MDCListFoundation extends MDCFoundation {
       }
     }
     this.adapter_.focusItemAtIndex(prevIndex);
+
+    return prevIndex;
   }
 
+  /**
+   * @return {number}
+   */
   focusFirstElement() {
     if (this.adapter_.getListItemCount() > 0) {
       this.adapter_.focusItemAtIndex(0);
+      return 0;
     }
   }
 
+  /**
+   * @return {number}
+   */
   focusLastElement() {
     const lastIndex = this.adapter_.getListItemCount() - 1;
     if (lastIndex >= 0) {
       this.adapter_.focusItemAtIndex(lastIndex);
+      return lastIndex;
     }
   }
 
+  /**
+   * @param {number} index
+   * @private
+   */
+  setSingleSelectionAtIndex_(index) {
+    let selectedClassName = cssClasses.LIST_ITEM_SELECTED_CLASS;
+    if (this.useActivatedClass_) {
+      selectedClassName = cssClasses.LIST_ITEM_ACTIVATED_CLASS;
+    }
+
+    if (this.selectedIndex_ >= 0 && this.selectedIndex_ !== index) {
+      this.adapter_.removeClassForElementIndex(this.selectedIndex_, selectedClassName);
+      this.adapter_.setAttributeForElementIndex(this.selectedIndex_, strings.ARIA_SELECTED, 'false');
+    }
+
+    this.adapter_.addClassForElementIndex(index, selectedClassName);
+    this.adapter_.setAttributeForElementIndex(index, strings.ARIA_SELECTED, 'true');
+
+    this.selectedIndex_ = index;
+  }
+
   /**
    * Toggles checkbox or radio at give index. Radio doesn't change the checked state if it is already checked.
    * @param {number} index
    * @private
    */
-  toggleCheckboxOrRadioAtIndex_(index) {
-    if (!this.hasCheckboxOrRadioAtIndex_(index)) return;
+  setRadioAtIndex_(index) {
+    this.adapter_.setCheckedCheckboxOrRadioAtIndex(index, true);
+
+    if (this.selectedIndex_ >= 0) {
+      this.adapter_.setAttributeForElementIndex(this.selectedIndex_, strings.ARIA_CHECKED, 'false');
+    }
+
+    this.adapter_.setAttributeForElementIndex(index, strings.ARIA_CHECKED, 'true');
 
-    let isChecked = true;
-    if (this.adapter_.hasCheckboxAtIndex(index)) {
-      isChecked = !this.adapter_.isCheckboxCheckedAtIndex(index);
+    this.selectedIndex_ = index;
+  }
+
+  /**
+   * @param {!Index} index
+   * @private
+   */
+  setCheckboxAtIndex_(index) {
+    if (this.programmaticSelection_) {
+      for (let i = 0; i < this.adapter_.getListItemCount(); i++) {
+        let isChecked = false;
+        if (index.indexOf(i) >= 0) {
+          isChecked = true;
+        }
+
+        this.adapter_.setCheckedCheckboxOrRadioAtIndex(i, isChecked);
+        this.adapter_.setAttributeForElementIndex(i, strings.ARIA_CHECKED, isChecked ? 'true' : 'false');
+      }
+
+      this.selectedIndex_ = index;
+    } else {
+      let isChecked = this.adapter_.isCheckboxCheckedAtIndex(index);
+      if (this.toggleCheckbox_) {
```

Comment:
I had the same idea of passing down the `toggleIndex` via selectedIndex private method so that it takes a different route.

But didn't want to expose optional parameter to public API that is meant for internal use. Instead, I do want to remove the need of passing `toggleIndex` and only send mouse event object similar to `handleKeydown` handler. Then the foundation can determine if it needs a toggle whether it is clicked on checkbox/radio input or on label. WDYT?

I merged above two conditions.